File: Data Structures & Algorithms/is-anagram/submission-0.java
class Solution {
    public boolean isAnagram(String s, String t) {
        HashMap<Character, Integer> freq = new HashMap<Character, Integer>();

        for (int i = 0; i < s.length(); i++) {
            Character c = s.charAt(i);
            Integer f = freq.get(c);
            if (f == null) {
                f = 1;
            } else {
                f++;
            }
            freq.put(c, f);
        }

        for (int i = 0; i < t.length(); i++) {
            Character c = t.charAt(i);
            Integer f = freq.get(c);
            if (f == null) {
                freq.put(c, -1);
            } else if (f == 1) {
                freq.remove(c);
            } else {
                freq.put(c, f - 1);
            }
        }

        return freq.isEmpty();
    }
}

File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-0.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        freq = {}

        sl = len(s)
        l = 0
        r = 0
        m = 0
        mw = 0

        for r in range(0, sl):
            c = s[r]
            n = freq.get(c, 0) + 1
            freq[c] = n

            if n > m:
                m = n

            if m < r - l + 1 - k:
                c = s[l]
            
                freq[c] = freq.get(c, 0) - 1

                l = l + 1

            if mw < r - l + 1:
                mw = r - l + 1

        return mw

